geomesa-gs-monitor-elasticsearch/src/main/scala/org/geomesa/gs/monitor/elastic/ElasticRequestDataListener.scala

```diff
@@ -8,137 +8,183 @@
+import org.springframework.context.{ApplicationEvent, ApplicationListener}
+
+import java.util.concurrent.{ConcurrentHashMap, Executors, LinkedBlockingQueue, TimeUnit}
+import java.util.function.BiFunction
+import scala.collection.JavaConverters._
+import scala.concurrent.TimeoutException
+
+class ElasticRequestDataListener extends RequestDataListener
+  with ApplicationListener[ApplicationEvent] with LazyLogging {
+
+  import org.geomesa.gs.monitor.elastic.ElasticRequestDataListener._
+
+  private val config = ConfigFactory.load.getConfig(GEOSERVER_MONITOR_ELASTICSEARCH_KEY)
+  private val index = config.getString("index")
+
+  private val excludedFields =
+    Option("excludedFields")
+      .filter(config.hasPath)
+      .map(config.getStringList)
+      .map(_.asScala.toSet)
+      .getOrElse(Set.empty)
+
+  private val timeout =
+    Option("timeout")
+      .filter(config.hasPath)
+      .map(config.getLong)
+      .filter { ms =>
+        if (ms >= 1000L && ms < 10000000L) {
+          true
+        } else {
+          logger.warn(s"Ignoring config value for timeout '$ms'")
+          false
+        }
+      }
+      .getOrElse(10000L)
+
+  private val gson = ExtendedRequestData.getGson(excludedFields)
+  private val restClient = getRestClient(config)
+
+  /*
+   * GeoServer Request Monitor Timeout Strategy
+   *
+   * If a request times out, it never reaches the completed or post-processed callbacks and thus
+   * would never be sent to Elasticsearch. To combat this, all in-progress requests are tracked.
+   * Whenever a request is post-processed, it is sent to Elasticsearch and removed from the set.
+   * Any in-progress requests that exceed the configured timeout are marked as failed and sent
+   * to Elasticsearch. If a request completes after it has already been marked as a timeout,
+   * the Elasticsearch index is updated using the request id.
+   */
+
+  private val writeQueue = new LinkedBlockingQueue[RequestData]
+  private val requestsInProgress = new ConcurrentHashMap[Long, RequestData]
+
+  private val elasticsearchWriter = new Runnable {
+    override def run(): Unit = {
+      while (!Thread.interrupted) {
+        val rd = writeQueue.take // block until data is available to send
+
+        try {
+          putElasticsearch(new ExtendedRequestData(rd))
+          logger.info(s"Sent request '${rd.uid}' to Elasticsearch")
+        } catch {
+          case ex: Exception => logger.error(s"Failed to send request '${rd.uid}' to Elasticsearch: ${ex.getMessage}")
+        }
+      }
+    }
+  }
 
-import scala.collection.mutable.ArrayBuffer
+  private val timeoutHandler = new Runnable {
+    override def run(): Unit = {
+      val currentMillis = System.currentTimeMillis
+
+      requestsInProgress.asScala
+        .filter { case (_, rd) =>
+          rd.getStartTimeMillis + timeout <= currentMillis
+        }
+        .foreach { case (id, _) =>
+          Option(requestsInProgress.remove(id)).foreach { request =>
```

Comment:
i think this could potentially still end up in the queue after the request completes unless you use the `compute` method like you did below.
i.e.
1. request times out and is removed from requestsInProgress
2. request completes and is added to the queue
3. foreach is executed and timed out request is added to the queue